File: packages/capacitor-notificare-monetize/android/src/main/java/re/notifica/monetize/capacitor/NotificareMonetizePlugin.kt
package re.notifica.monetize.capacitor

import android.os.Handler
import android.os.Looper
import androidx.lifecycle.Observer
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.internal.NotificareLogger
import re.notifica.monetize.NotificareMonetize
import re.notifica.monetize.ktx.monetize
import re.notifica.monetize.models.NotificareProduct
import re.notifica.monetize.models.NotificarePurchase

@CapacitorPlugin(name = "NotificareMonetizePlugin")
public class NotificareMonetizePlugin : Plugin(), NotificareMonetize.Listener {
    private val productsObserver = Observer<List<NotificareProduct>> { products ->
        if (products == null) return@Observer

        try {
            val data = JSObject()
            data.put("products", JSArray(products.map { it.toJson() }))

            EventBroker.dispatchEvent("products_updated", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the products_updated event.", e)
        }
    }

    private val purchasesObserver = Observer<List<NotificarePurchase>> { purchases ->
        if (purchases == null) return@Observer

        try {
            val data = JSObject()
            data.put("purchases", JSArray(purchases.map { it.toJson() }))

            EventBroker.dispatchEvent("purchases_updated", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchases_updated event.", e)
        }
    }

    override fun load() {
        EventBroker.setup(this::notifyListeners)

        onMainThread {
            Notificare.monetize().observableProducts.observeForever(productsObserver)
            Notificare.monetize().observablePurchases.observeForever(purchasesObserver)
        }

        Notificare.monetize().removeListener(this)
        Notificare.monetize().addListener(this)
    }

    // region Notificare Monetize

    @PluginMethod
    public fun getProducts(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", JSArray(Notificare.monetize().products.map { it.toJson() }))
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun getPurchases(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", JSArray(Notificare.monetize().purchases.map { it.toJson() }))
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun refresh(call: PluginCall) {
        Notificare.monetize().refresh(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun startPurchaseFlow(call: PluginCall) {
        val json = call.getObject("product") ?: run {
            call.reject("Missing 'product' parameter.")
            return
        }

        val product: NotificareProduct = try {
            NotificareProduct.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        val activity = activity ?: run {
            call.reject("Cannot start a purchase without an activity.")
            return
        }

        Notificare.monetize().startPurchaseFlow(activity, product)
        call.resolve()
    }

    // region NotificareMonetize.Listener

    override fun onBillingSetupFinished() {
        try {
            EventBroker.dispatchEvent("billing_setup_finished", null)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the billing_setup_failed event.", e)
        }
    }

    override fun onBillingSetupFailed(code: Int, message: String) {
        try {
            val data = JSObject().apply {
                put("code", code)
                put("message", message)
            }

            EventBroker.dispatchEvent("billing_setup_failed", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the billing_setup_failed event.", e)
        }
    }

    override fun onPurchaseFinished(purchase: NotificarePurchase) {
        try {
            EventBroker.dispatchEvent("purchase_finished", purchase.toJson())
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_finished event.", e)
        }
    }

    override fun onPurchaseRestored(purchase: NotificarePurchase) {
        try {
            EventBroker.dispatchEvent("purchase_restored", purchase.toJson())
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_restored event.", e)
        }
    }

    override fun onPurchaseCanceled() {
        try {
            EventBroker.dispatchEvent("purchase_canceled", null)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_canceled event.", e)
        }
    }

    override fun onPurchaseFailed(code: Int, message: String) {
        try {
            val data = JSObject().apply {
                put("code", code)
                put("message", message)
            }

            EventBroker.dispatchEvent("purchase_failed", data)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_failed event.", e)
        }
    }

    // endregion

    // endregion

    public companion object {
        internal fun onMainThread(action: () -> Unit) = Handler(Looper.getMainLooper()).post(action)
    }
}

File: packages/capacitor-notificare-monetize/ios/Plugin/NotificareMonetizePlugin.swift
import Foundation
import Capacitor
import NotificareKit
import NotificareMonetizeKit

@objc(NotificareMonetizePlugin)
public class NotificareMonetizePlugin: CAPPlugin {
    public override func load() {
        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Notificare.shared.monetize().delegate = self
    }
    
    @objc func getProducts(_ call: CAPPluginCall) {
        do {
            call.resolve([
                "result": try Notificare.shared.monetize().products.map { try $0.toJson() }
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func getPurchases(_ call: CAPPluginCall) {
        do {
            call.resolve([
                "result": try Notificare.shared.monetize().purchases.map { try $0.toJson() }
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func refresh(_ call: CAPPluginCall) {
        Notificare.shared.monetize().refresh { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func startPurchaseFlow(_ call: CAPPluginCall) {
        guard let json = call.getObject("product") else {
            call.reject("Missing 'product' parameter.")
            return
        }
        
        let product: NotificareProduct
        
        do {
            product = try NotificareProduct.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.monetize().startPurchaseFlow(for: product)
        call.resolve()
    }
}

extension NotificareMonetizePlugin: NotificareMonetizeDelegate {
    public func notificare(_ notificareMonetize: NotificareMonetize, didUpdateProducts products: [NotificareProduct]) {
        do {
            let data = [
                "products": try products.map { try $0.toJson() }
            ]
            
            EventBroker.instance.dispatchEvent("products_updated", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the products_updated event.", error: error)
        }
    }
    
    public func notificare(_ notificareMonetize: NotificareMonetize, didUpdatePurchases purchases: [NotificarePurchase]) {
        do {
            let data = [
                "purchases": try purchases.map { try $0.toJson() }
            ]
            
            EventBroker.instance.dispatchEvent("purchases_updated", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the purchases_updated event.", error: error)
        }
    }
    
    public func notificare(_ notificareMonetize: NotificareMonetize, didFinishPurchase purchase: NotificarePurchase) {
        do {
            EventBroker.instance.dispatchEvent("purchase_finished", data: try purchase.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the purchase_finished event.", error: error)
        }
    }
    
    public func notificare(_ notificareMonetize: NotificareMonetize, didRestorePurchase purchase: NotificarePurchase) {
        do {
            EventBroker.instance.dispatchEvent("purchase_restored", data: try purchase.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the purchase_restored event.", error: error)
        }
    }
    
    public func notificareDidCancelPurchase(_ notificareMonetize: NotificareMonetize) {
        EventBroker.instance.dispatchEvent("purchase_canceled", data: nil)
    }
    
    public func notificare(_ notificareMonetize: NotificareMonetize, didFailToPurchase error: Error) {
        EventBroker.instance.dispatchEvent("purchase_failed", data: ["errorMessage": error.localizedDescription])
    }
}
